contracts/contracts/GrantRoundManager.sol

```diff
@@ -0,0 +1,171 @@
+
+    registry = _registry;
+    router = _router;
+    donationToken = _donationToken;
+  }
+
+  /**
+   * @notice Creates a new GrantRound
+   * @param _owner Grant round owner that has permission to update the metadata pointer
+   * @param _payoutAdmin Grant round administrator that has permission to payout the matching pool
+   * @param _registry Address that contains the grant metadata
+   * @param _startTime Unix timestamp of the start of the round
+   * @param _endTime Unix timestamp of the end of the round
+   * @param _metaPtr URL pointing to the grant round metadata
+   * @param _minContribution Miniumum donation amount that can be made using the given donation token
+   */
+  function createGrantRound(
+    address _owner,
+    address _payoutAdmin,
+    GrantRegistry _registry,
+    uint256 _startTime,
+    uint256 _endTime,
+    string memory _metaPtr,
+    uint256 _minContribution
+  ) external {
+    GrantRound _grantRound = new GrantRound(
+      _owner,
+      _payoutAdmin,
+      _registry,
+      donationToken,
+      _startTime,
+      _endTime,
+      _metaPtr,
+      _minContribution
+    );
+
+    emit GrantRoundCreated(address(_grantRound));
+  }
+
+  /**
+   * @notice Swap and donate to a grant
+   * @param  _donation Donation being made to a grant
+   */
+  function swapAndDonate(Donation calldata _donation) external payable {
+    // --- Validation ---
+    // Rounds must be specified
+    GrantRound[] calldata _rounds = _donation.rounds;
+    require(_rounds.length > 0, "GrantRoundManager: Must specify at least one round");
+
+    // Only allow value to be sent if the input token is WETH
+    IERC20 _tokenIn = _donation.tokenIn;
+    require(
+      (msg.value == 0 && address(_tokenIn) != WETH) || (msg.value > 0 && address(_tokenIn) == WETH),
+      "GrantRoundManager: Invalid token-value pairing"
+    );
+
+    // Wnsure grant recieving donation exists in registry
+    uint96 _grantId = _donation.grantId;
+    require(_grantId < registry.grantCount(), "GrantRoundManager: Grant does not exist in registry");
+
+    // Iterate through each GrantRound to verify:
+    //   - The round has the same donationToken as the GrantRoundManager
+    //   - The round is active
+    for (uint256 i = 0; i < _rounds.length; i++) {
+      require(
+        donationToken == _rounds[i].donationToken(),
+        "GrantRoundManager: GrantRound's donation token does not match GrantRoundManager's donation token"
+      );
+
+      require(
+        block.timestamp >= _rounds[i].startTime() && block.timestamp < _rounds[i].endTime(),
+        "GrantRoundManager: GrantRound is not active"
+      );
+    }
+
+    // --- Swap ---
+    address _payoutAddress = registry.getGrantPayee(_grantId);
+    uint256 _amountIn = _donation.amountIn;
+    uint256 _amountOut = _amountIn; // by default, by may be overwritten in the swap branch below
+
+    if (_tokenIn == donationToken && msg.value == 0) {
```

Comment:
Unclear what checking for `msg.value == 0` accomplishes. If it's a sanity check, then it would appear we're missing the corresponding sanity check in the `else` branch, which seems to assume without validating— if I'm reading it right— that if there `msg.value` being sent, then `_tokenIn` must be the 0xEeE address for ETH.

Reply:
We don't support a `donationToken` of ETH, so we want to make sure if ETH is sent we don't attempt a transfer.

The missing branch would be added if/when we support ETH, and would do an ETH transfer instead of `_tokenIn.safeTransferFrom()`. That would also require the special ETH_ADDRESS to be defined which we don't have, so I figured we don't need that branch yet.

> we're missing the corresponding sanity check in the `else` branch

Sorry, which check are we missing? In that branch, we just need `_tokenIn != donationToken` and it supports any `msg.value`, so it seems covered

Reply:
I guess we could revert if `_tokenIn == donationToken && msg.value > 0` 🤔

Reply:
Plus this will all get modified soon anyway, so not too worried about it just yet!

Reply:
> I guess we could revert if _tokenIn == donationToken && msg.value > 0, but Uniswap will fail anyway in that case. Need to check if it fails with a meaningful message

Didn't realize Uniswap would fail. We should at least document that with a comment or the like.

> Plus this will all get modified soon anyway, so not too worried about it just yet!

Agreed.